tips/030_speculative_calls.md

```diff
@@ -0,0 +1,217 @@
+## Effective AI Engineering #30: Speculative Calls
+
+**Are your users waiting seconds for simple yes/no decisions?** Your AI classifies user intent, then routes to the appropriate handler, but that sequential flow adds unnecessary latency to your hot path.
+
+When AI workflows involve classification followed by specialized processing, the linear approach forces users to wait for each step. But if you can predict the likely outcome, you can start expensive work early and provide instant responses.
+
+### The Problem
+
+Many developers implement AI workflows as strict sequences, waiting for each step before starting the next. This creates challenges that aren't immediately obvious:
+
+```python
+# BEFORE: Sequential processing with classification bottleneck
+from mirascope.core import llm
+import lilypad
+import time
+
+@lilypad.trace()
+@llm.call(provider="openai", model="gpt-4o-mini")
+def classify_request(user_request: str) -> str:
+    return f"Classify this request as 'simple' or 'complex': {user_request}"
+
+@lilypad.trace()
+@llm.call(provider="openai", model="gpt-4o-mini")
+def handle_simple_request(user_request: str) -> str:
+    return f"Handle this simple request: {user_request}"
+
+@lilypad.trace()
+@llm.call(provider="openai", model="gpt-4o-mini")
+def handle_complex_request(user_request: str) -> str:
+    return f"Handle this complex request with detailed analysis: {user_request}"
+
+@lilypad.trace()
+def process_user_request(user_request: str) -> str:
+    start_time = time.time()
+    
+    # Step 1: Classify (adds latency)
+    classification = classify_request(user_request)
+    print(f"Classification took: {time.time() - start_time:.2f}s")
+    
+    # Step 2: Route based on classification
+    if classification == "simple":
+        result = handle_simple_request(user_request)
+    else:
+        result = handle_complex_request(user_request)
+    
+    print(f"Total time: {time.time() - start_time:.2f}s")
+    return result
+
+# Sequential approach - classification blocks processing
+result = process_user_request("What's the weather like?")
+```
+
+**Why this approach falls short:**
+
+- **Sequential Latency:** Users wait for classification before any real work begins
+- **Underutilized Resources:** System idles during classification instead of doing useful work
+- **Poor User Experience:** Simple requests feel slow despite having fast answers
+
+### The Solution: Parallel Speculative Execution
+
+A better approach is to predict likely outcomes and start work speculatively in parallel with classification. This pattern optimizes for the common case while gracefully handling mispredictions.
+
+```python
+# AFTER: Speculative parallel processing
+from mirascope.core import llm
```

Comment:
this AFTER exacmple is way too complicated. Needs to be simplified. Just do the most straightforward thing. We don't need confidence and other shit.